controller/rancher/certificate_fetcher.go

```
@@ -0,0 +1,317 @@
+	if f != nil && f.IsDir() {
+		logrus.Debugf("walking %v", path)
+		isCertFound := false
+		cert := config.Certificate{}
+		cert.Name = f.Name()
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if !file.IsDir() {
+				contentBytes, err := fetcher.evaluatueLinkAndReadFile(path, file.Name())
+				if err != nil {
+					return err
+				}
+				if file.Name() == CertName {
+					isCertFound = true
+					cert.Cert = string(*contentBytes)
+				} else if file.Name() == KeyName {
+					isCertFound = true
+					cert.Key = string(*contentBytes)
+				}
+			}
+		}
+		if isCertFound {
+			fetcher.TempCertsMap[path] = &cert
+		}
+	}
+	return nil
+}
+
+func (fetcher *RCertificateFetcher) evaluatueLinkAndReadFile(relativePath string, fileName string) (*[]byte, error) {
+	filePath := path.Join(relativePath, fileName)
+
+	absFilePath, err := filepath.Abs(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error forming path to file %s, error: %v", filePath, err)
+	}
+
+	fInfo, err := os.Lstat(absFilePath)
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("File %s does not exist", absFilePath)
+	}
+
+	targetPath := absFilePath
+	if fInfo.Mode()&os.ModeSymlink != 0 {
+		//read symlink
+		targetPath, err := filepath.EvalSymlinks(absFilePath)
+		if err != nil {
+			return nil, fmt.Errorf("File %s pointed by symlink %s does not exist, error: %v", targetPath, absFilePath, err)
+		}
+	}
+	//read target file
+	return fetcher.readFile(targetPath)
+}
+
+func (fetcher *RCertificateFetcher) readFile(filePath string) (*[]byte, error) {
+	contentBytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading file %s, error: %v", filePath, err)
+	}
+	return &contentBytes, nil
+}
+
+func (fetcher *RCertificateFetcher) addWatches() {
+	if fetcher.CertDir != "" {
+		err := fetcher.Watcher.Add(fetcher.CertDir)
+		if err != nil {
+			logrus.Errorf("Failed to add watch for Dir %v", fetcher.CertDir)
```

We should probably return an error when addWatches fail.